homeassistant/components/geniushub/__init__.py

```diff
@@ -19,36 +27,62 @@
 )
 from homeassistant.helpers.entity import Entity
 from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+import homeassistant.util.dt as dt_util
+
+ATTR_DURATION = "duration"
 
 _LOGGER = logging.getLogger(__name__)
 
 DOMAIN = "geniushub"
 
+# temperature is repeated here, as it gives access to high-precision temps
+GH_ZONE_ATTRS = ["mode", "temperature", "type", "occupied", "override"]
+GH_DEVICE_ATTRS = {
+    "luminance": "luminance",
+    "measuredTemperature": "measured_temperature",
+    "occupancyTrigger": "occupancy_trigger",
+    "setback": "setback",
+    "setTemperature": "set_temperature",
+    "wakeupInterval": "wakeup_interval",
+}
+
 SCAN_INTERVAL = timedelta(seconds=60)
 
-_V1_API_SCHEMA = vol.Schema({vol.Required(CONF_TOKEN): cv.string})
-_V3_API_SCHEMA = vol.Schema(
+MAC_ADDRESS_REGEXP = r"^([0-9A-F]{2}:){5}([0-9A-F]{2})$"
+
+V1_API_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_TOKEN): cv.string,
+        vol.Required(CONF_MAC): vol.Match(MAC_ADDRESS_REGEXP),
+    }
+)
+V3_API_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_HOST): cv.string,
         vol.Required(CONF_USERNAME): cv.string,
         vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_MAC): vol.Match(MAC_ADDRESS_REGEXP),
     }
 )
 CONFIG_SCHEMA = vol.Schema(
-    {DOMAIN: vol.Any(_V3_API_SCHEMA, _V1_API_SCHEMA)}, extra=vol.ALLOW_EXTRA
+    {DOMAIN: vol.Any(V3_API_SCHEMA, V1_API_SCHEMA)}, extra=vol.ALLOW_EXTRA
 )
 
 
-async def async_setup(hass, hass_config):
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
     """Create a Genius Hub system."""
-    kwargs = dict(hass_config[DOMAIN])
+    hass.data[DOMAIN] = {}
+
+    kwargs = dict(config[DOMAIN])
     if CONF_HOST in kwargs:
         args = (kwargs.pop(CONF_HOST),)
     else:
         args = (kwargs.pop(CONF_TOKEN),)
 
-    hass.data[DOMAIN] = {}
-    broker = GeniusBroker(hass, args, kwargs)
+    broker = hass.data[DOMAIN]["broker"] = GeniusBroker(hass, args, kwargs)
+
+    broker.client = GeniusHub(*args, **kwargs, session=async_get_clientsession(hass))
```

Comment:
Pass the client as an argument to the broker when creating the broker.

Reply:
Done.

@MartinHjelmare, I swear you're getting more strict as time goes on!